hazelcast/src/main/java/com/hazelcast/nio/tcp/nonblocking/NonBlockingIOThread.java

```diff
@@ -87,22 +104,85 @@ public final void addTask(Runnable task) {
      * eventually pick up the task.
      *
      * @param task the task to add.
+     * @throws NullPointerException if task is null
      */
     public final void addTaskAndWakeup(Runnable task) {
-        selectorQueue.add(task);
+        taskQueue.add(task);
         selector.wakeup();
     }
 
-    // shows how long this probe has been idle.
-    @Probe
-    private long idleTime() {
-        return Math.max(System.currentTimeMillis() - lastSelectTimeMs, 0);
+    @Override
+    public final void run() {
+        // This outer loop is a bit complex but it takes care of a lot of stuff:
+        // * it calls runSelectNowLoop or runSelectLoop based on selectNow enabled or not.
+        // * handles backoff and retrying in case if io exception is thrown
+        // * it takes care of other exception handling.
+        //
+        // The idea about this approach is that the runSelectNowLoop and runSelectLoop are as clean as possible and don't contain
+        // any logic that isn't happening on the happy-path.
+        try {
+            for (; ; ) {
+                try {
+                    if (selectNow) {
```

Comment:
this is ugly, but as already discussed: I don't know any better so I can live with it.